lib/syncthing/auditservice.go

```diff
@@ -38,7 +38,10 @@ func (s *auditService) Serve(ctx context.Context) error {
 
 	for {
 		select {
-		case ev := <-sub.C():
+		case ev, ok := <-sub.C():
+			if !ok {
+				return ctx.Err()
```

Comment:
I wonder if we're guaranteed that the context is in fact closed at this point, or if it might be better to store `sub.C()` in a variable and nil it out when it closes, so that we just wait for `ctx.Done()` on the next loop?

Reply:
The current implementation seems to guarantee it - the only place outside of Subscribe that closes the `events` channel is in `Serve`, on `context.Done()`.

I'm not sure I understand your suggested alternative solution - do you mean another goroutine that nils the channel variable? If so, that would still have a potential race AFAICT, though a much slimmer one. The race would be between the goroutine that nils the local variable upon close and the main loop which tries to read from it in a select. I might be missing something though :)

Reply:
What I meant was something like

```
events := sub.C()
for {
  select {
    case ev, ok := <-events:
      if !ok { events = nil }
      ...
```

but now I'm thinking whether it's useful to even close that channel at all to begin with. Probably there is no code that takes advantage of the event channel closing when the subscription does...

Reply:
Maybe add a wait on the context just to be sure?

```suggestion
				<-ctx.Done()
				return ctx.Err()
```

Reply:
That sounds reasonable

Reply:
@calmh I was wondering about removing the channel close, but it does affect places that don't wait on the context such as the Poll method.

I see what you mean now - I think that makes sense. Would you prefer that solution or the one @greatroar suggested?

Reply:
Waiting for the context is less code; let's do that.